#!/usr/bin/env python3
"""
å®Œæ•´çš„ MinIO gRPC è¿œç¨‹è®¿é—®å®¢æˆ·ç«¯
æ–‡ä»¶å: test_minio_remote_client.py

è¿™ä¸ªè„šæœ¬æ¼”ç¤ºå¦‚ä½•ä»è¿œç¨‹æœåŠ¡è®¿é—® MinIO gRPC æœåŠ¡ï¼Œå¹¶å®é™…è°ƒç”¨æœåŠ¡æ–¹æ³•ã€‚
"""

import grpc
import sys
from datetime import datetime

# å¯¼å…¥ç”Ÿæˆçš„ proto æ–‡ä»¶
sys.path.insert(0, '/Users/xenodennis/Documents/Fun/isA_user/tests')
from proto import minio_service_pb2
from proto import minio_service_pb2_grpc


class MinIORemoteClient:
    """MinIO gRPC è¿œç¨‹å®¢æˆ·ç«¯"""
    
    def __init__(self, host='localhost', port=50051):
        """
        åˆå§‹åŒ–å®¢æˆ·ç«¯
        
        Args:
            host: æœåŠ¡å™¨åœ°å€
                  - æœ¬åœ°: 'localhost'
                  - å±€åŸŸç½‘: '192.168.31.62'
                  - å…¬ç½‘: 'your-server.com'
            port: æœåŠ¡ç«¯å£ï¼ˆé»˜è®¤ 50051ï¼‰
        
        ç¤ºä¾‹:
            # æœ¬åœ°è®¿é—®
            client = MinIORemoteClient('localhost', 50051)
            
            # è¿œç¨‹è®¿é—®
            client = MinIORemoteClient('192.168.31.62', 50051)
        """
        self.host = host
        self.port = port
        self.address = f'{host}:{port}'
        
        print(f"ğŸ”— è¿æ¥åˆ° MinIO gRPC æœåŠ¡: {self.address}")
        
        # åˆ›å»º gRPC channel
        # æ³¨æ„: ç”Ÿäº§ç¯å¢ƒåº”è¯¥ä½¿ç”¨ secure_channel
        self.channel = grpc.insecure_channel(
            self.address,
            options=[
                ('grpc.max_receive_message_length', 100 * 1024 * 1024),  # 100MB
                ('grpc.max_send_message_length', 100 * 1024 * 1024),     # 100MB
                ('grpc.keepalive_time_ms', 10000),
                ('grpc.keepalive_timeout_ms', 5000),
            ]
        )
        
        # åˆ›å»º stub
        self.stub = minio_service_pb2_grpc.MinIOServiceStub(self.channel)
        
        # æµ‹è¯•è¿æ¥
        try:
            grpc.channel_ready_future(self.channel).result(timeout=5)
            print(f"âœ… è¿æ¥æˆåŠŸ!\n")
        except Exception as e:
            print(f"âŒ è¿æ¥å¤±è´¥: {e}\n")
            raise
    
    def health_check(self, detailed=True):
        """
        å¥åº·æ£€æŸ¥
        
        Args:
            detailed: æ˜¯å¦è¿”å›è¯¦ç»†ä¿¡æ¯
        
        Returns:
            HealthCheckResponse
        """
        print("=" * 60)
        print("æµ‹è¯• 1: å¥åº·æ£€æŸ¥ (HealthCheck)")
        print("=" * 60)
        
        try:
            request = minio_service_pb2.HealthCheckRequest(detailed=detailed)
            response = self.stub.HealthCheck(request)
            
            print(f"âœ… æœåŠ¡çŠ¶æ€: {response.status}")
            print(f"   å¥åº·: {response.healthy}")
            print(f"   æˆåŠŸ: {response.success}")
            if response.details:
                print(f"   è¯¦ç»†ä¿¡æ¯: {dict(response.details)}")
            
            return response
            
        except grpc.RpcError as e:
            print(f"âŒ RPC é”™è¯¯: {e.code()}")
            print(f"   è¯¦æƒ…: {e.details()}")
            return None
    
    def create_bucket(self, bucket_name, user_id, organization_id='remote-org'):
        """
        åˆ›å»ºå­˜å‚¨æ¡¶
        
        Args:
            bucket_name: æ¡¶åç§°
            user_id: ç”¨æˆ· ID
            organization_id: ç»„ç»‡ ID
        
        Returns:
            CreateBucketResponse
        """
        print("\n" + "=" * 60)
        print(f"æµ‹è¯• 2: åˆ›å»ºå­˜å‚¨æ¡¶ (CreateBucket)")
        print("=" * 60)
        print(f"æ¡¶åç§°: {bucket_name}")
        print(f"ç”¨æˆ· ID: {user_id}")
        print(f"ç»„ç»‡ ID: {organization_id}")
        
        try:
            request = minio_service_pb2.CreateBucketRequest(
                bucket_name=bucket_name,
                user_id=user_id,
                organization_id=organization_id,
                region='us-east-1'
            )
            
            response = self.stub.CreateBucket(request)
            
            if response.success:
                print(f"âœ… æ¡¶åˆ›å»ºæˆåŠŸ!")
                if response.bucket_info:
                    print(f"   åç§°: {response.bucket_info.name}")
                    print(f"   æ‰€æœ‰è€…: {response.bucket_info.owner_id}")
                    print(f"   ç»„ç»‡: {response.bucket_info.organization_id}")
            else:
                print(f"âš ï¸  {response.message or response.error}")
            
            return response
            
        except grpc.RpcError as e:
            print(f"âŒ RPC é”™è¯¯: {e.code()}")
            print(f"   è¯¦æƒ…: {e.details()}")
            return None
    
    def list_buckets(self, user_id, organization_id='remote-org'):
        """
        åˆ—å‡ºå­˜å‚¨æ¡¶
        
        Args:
            user_id: ç”¨æˆ· ID
            organization_id: ç»„ç»‡ ID
        
        Returns:
            ListBucketsResponse
        """
        print("\n" + "=" * 60)
        print(f"æµ‹è¯• 3: åˆ—å‡ºå­˜å‚¨æ¡¶ (ListBuckets)")
        print("=" * 60)
        
        try:
            request = minio_service_pb2.ListBucketsRequest(
                user_id=user_id,
                organization_id=organization_id
            )
            
            response = self.stub.ListBuckets(request)
            
            if response.success:
                print(f"âœ… æ‰¾åˆ° {len(response.buckets)} ä¸ªæ¡¶:")
                for i, bucket in enumerate(response.buckets, 1):
                    print(f"   {i}. {bucket.name}")
                    print(f"      æ‰€æœ‰è€…: {bucket.owner_id}")
                    print(f"      ç»„ç»‡: {bucket.organization_id}")
            else:
                print(f"âš ï¸  {response.error}")
            
            return response
            
        except grpc.RpcError as e:
            print(f"âŒ RPC é”™è¯¯: {e.code()}")
            print(f"   è¯¦æƒ…: {e.details()}")
            return None
    
    def upload_object(self, bucket_name, object_key, data, user_id, content_type='text/plain'):
        """
        ä¸Šä¼ å¯¹è±¡ï¼ˆæµå¼ï¼‰
        
        Args:
            bucket_name: æ¡¶åç§°
            object_key: å¯¹è±¡é”®ï¼ˆè·¯å¾„ï¼‰
            data: äºŒè¿›åˆ¶æ•°æ®
            user_id: ç”¨æˆ· ID
            content_type: å†…å®¹ç±»å‹
        
        Returns:
            PutObjectResponse
        """
        print("\n" + "=" * 60)
        print(f"æµ‹è¯• 4: ä¸Šä¼ å¯¹è±¡ (PutObject - æµå¼)")
        print("=" * 60)
        print(f"æ¡¶: {bucket_name}")
        print(f"å¯¹è±¡é”®: {object_key}")
        print(f"æ•°æ®å¤§å°: {len(data)} bytes")
        print(f"å†…å®¹ç±»å‹: {content_type}")
        
        try:
            def request_generator():
                # ç¬¬ä¸€ä¸ªæ¶ˆæ¯ï¼šå…ƒæ•°æ®
                metadata = minio_service_pb2.PutObjectMetadata(
                    bucket_name=bucket_name,
                    object_key=object_key,
                    user_id=user_id,
                    content_type=content_type,
                    content_length=len(data)
                )
                yield minio_service_pb2.PutObjectRequest(metadata=metadata)
                
                # åç»­æ¶ˆæ¯ï¼šæ•°æ®å—
                chunk_size = 1024 * 64  # 64KB chunks
                for i in range(0, len(data), chunk_size):
                    chunk = data[i:i + chunk_size]
                    yield minio_service_pb2.PutObjectRequest(chunk=chunk)
                    print(f"   ä¸Šä¼ è¿›åº¦: {min(i + chunk_size, len(data))}/{len(data)} bytes")
            
            response = self.stub.PutObject(request_generator())
            
            if response.success:
                print(f"âœ… å¯¹è±¡ä¸Šä¼ æˆåŠŸ!")
                print(f"   å¯¹è±¡é”®: {response.object_key}")
                print(f"   å¤§å°: {response.size} bytes")
                print(f"   ETag: {response.etag}")
            else:
                print(f"âš ï¸  {response.error}")
            
            return response
            
        except grpc.RpcError as e:
            print(f"âŒ RPC é”™è¯¯: {e.code()}")
            print(f"   è¯¦æƒ…: {e.details()}")
            return None
    
    def list_objects(self, bucket_name, user_id, prefix=''):
        """
        åˆ—å‡ºå¯¹è±¡
        
        Args:
            bucket_name: æ¡¶åç§°
            user_id: ç”¨æˆ· ID
            prefix: å‰ç¼€è¿‡æ»¤
        
        Returns:
            ListObjectsResponse
        """
        print("\n" + "=" * 60)
        print(f"æµ‹è¯• 5: åˆ—å‡ºå¯¹è±¡ (ListObjects)")
        print("=" * 60)
        print(f"æ¡¶: {bucket_name}")
        
        try:
            request = minio_service_pb2.ListObjectsRequest(
                bucket_name=bucket_name,
                user_id=user_id,
                prefix=prefix,
                max_keys=100
            )
            
            response = self.stub.ListObjects(request)
            
            if response.success:
                print(f"âœ… æ‰¾åˆ° {len(response.objects)} ä¸ªå¯¹è±¡:")
                for i, obj in enumerate(response.objects, 1):
                    print(f"   {i}. {obj.key}")
                    print(f"      å¤§å°: {obj.size} bytes")
                    print(f"      ç±»å‹: {obj.content_type}")
                    print(f"      ETag: {obj.etag}")
            else:
                print(f"âš ï¸  {response.error}")
            
            return response
            
        except grpc.RpcError as e:
            print(f"âŒ RPC é”™è¯¯: {e.code()}")
            print(f"   è¯¦æƒ…: {e.details()}")
            return None
    
    def get_presigned_url(self, bucket_name, object_key, user_id, expiry_seconds=3600):
        """
        è·å–é¢„ç­¾å URL
        
        Args:
            bucket_name: æ¡¶åç§°
            object_key: å¯¹è±¡é”®
            user_id: ç”¨æˆ· ID
            expiry_seconds: è¿‡æœŸæ—¶é—´ï¼ˆç§’ï¼‰
        
        Returns:
            GetPresignedURLResponse
        """
        print("\n" + "=" * 60)
        print(f"æµ‹è¯• 6: è·å–é¢„ç­¾å URL (GetPresignedURL)")
        print("=" * 60)
        
        try:
            request = minio_service_pb2.GetPresignedURLRequest(
                bucket_name=bucket_name,
                object_key=object_key,
                user_id=user_id,
                expiry_seconds=expiry_seconds
            )
            
            response = self.stub.GetPresignedURL(request)
            
            if response.success:
                print(f"âœ… é¢„ç­¾å URL ç”ŸæˆæˆåŠŸ!")
                print(f"   URL: {response.url[:80]}...")
                print(f"   è¿‡æœŸæ—¶é—´: {response.expires_at}")
            else:
                print(f"âš ï¸  {response.error}")
            
            return response
            
        except grpc.RpcError as e:
            print(f"âŒ RPC é”™è¯¯: {e.code()}")
            print(f"   è¯¦æƒ…: {e.details()}")
            return None
    
    def close(self):
        """å…³é—­è¿æ¥"""
        self.channel.close()
        print("\n" + "=" * 60)
        print("ğŸ”Œ è¿æ¥å·²å…³é—­")
        print("=" * 60)


def main():
    """ä¸»æµ‹è¯•å‡½æ•°"""
    print("=" * 60)
    print("  MinIO gRPC è¿œç¨‹è®¿é—®å®Œæ•´æµ‹è¯•")
    print("=" * 60)
    print()
    print("ğŸ“ æœåŠ¡ä¿¡æ¯:")
    print("   å®¹å™¨: isa-minio-grpc")
    print("   ç«¯å£æ˜ å°„: 0.0.0.0:50051->50051/tcp")
    print()
    print("ğŸŒ æ”¯æŒçš„è®¿é—®æ–¹å¼:")
    print("   1. æœ¬åœ°è®¿é—®: localhost:50051")
    print("   2. å±€åŸŸç½‘è®¿é—®: 192.168.31.62:50051")
    print("   3. å…¬ç½‘è®¿é—®: <å…¬ç½‘IP>:50051 (éœ€é…ç½®)")
    print()
    
    # è¯¢é—®ä½¿ç”¨å“ªç§è®¿é—®æ–¹å¼
    print("è¯·é€‰æ‹©è®¿é—®æ–¹å¼:")
    print("  1 - æœ¬åœ°è®¿é—® (localhost)")
    print("  2 - å±€åŸŸç½‘è¿œç¨‹è®¿é—® (192.168.31.62)")
    print("  3 - è‡ªå®šä¹‰åœ°å€")
    
    choice = input("\né€‰æ‹© [1]: ").strip() or "1"
    
    if choice == "1":
        host = 'localhost'
    elif choice == "2":
        host = '192.168.31.62'
    elif choice == "3":
        host = input("è¯·è¾“å…¥æœåŠ¡å™¨åœ°å€: ").strip()
    else:
        host = 'localhost'
    
    print()
    
    try:
        # åˆ›å»ºå®¢æˆ·ç«¯
        client = MinIORemoteClient(host=host, port=50051)
        
        # æµ‹è¯•å‚æ•°
        test_user_id = f'remote-user-{datetime.now().strftime("%Y%m%d%H%M%S")}'
        test_bucket_name = f'remote-test-bucket-{datetime.now().strftime("%H%M%S")}'
        test_object_key = 'test-file.txt'
        test_data = f"Hello from remote gRPC client!\nTimestamp: {datetime.now()}\nHost: {host}\n".encode()
        
        print(f"ğŸ“‹ æµ‹è¯•å‚æ•°:")
        print(f"   ç”¨æˆ· ID: {test_user_id}")
        print(f"   æ¡¶åç§°: {test_bucket_name}")
        print(f"   å¯¹è±¡é”®: {test_object_key}")
        print()
        
        # æ‰§è¡Œæµ‹è¯•
        # 1. å¥åº·æ£€æŸ¥
        client.health_check(detailed=True)
        
        # 2. åˆ›å»ºæ¡¶
        client.create_bucket(
            bucket_name=test_bucket_name,
            user_id=test_user_id,
            organization_id='remote-test-org'
        )
        
        # 3. åˆ—å‡ºæ¡¶
        client.list_buckets(user_id=test_user_id)
        
        # 4. ä¸Šä¼ å¯¹è±¡
        client.upload_object(
            bucket_name=test_bucket_name,
            object_key=test_object_key,
            data=test_data,
            user_id=test_user_id,
            content_type='text/plain'
        )
        
        # 5. åˆ—å‡ºå¯¹è±¡
        client.list_objects(
            bucket_name=test_bucket_name,
            user_id=test_user_id
        )
        
        # 6. è·å–é¢„ç­¾å URL
        client.get_presigned_url(
            bucket_name=test_bucket_name,
            object_key=test_object_key,
            user_id=test_user_id,
            expiry_seconds=3600
        )
        
        # å…³é—­è¿æ¥
        client.close()
        
        print()
        print("=" * 60)
        print("âœ… æ‰€æœ‰æµ‹è¯•å®Œæˆ!")
        print("=" * 60)
        print()
        print("ğŸ’¡ ä»£ç ç¤ºä¾‹ (åœ¨å…¶ä»–é¡¹ç›®ä¸­ä½¿ç”¨):")
        print()
        print("```python")
        print("import grpc")
        print("from proto import minio_service_pb2, minio_service_pb2_grpc")
        print()
        print(f"# è¿æ¥åˆ°è¿œç¨‹æœåŠ¡")
        print(f"channel = grpc.insecure_channel('{host}:50051')")
        print("stub = minio_service_pb2_grpc.MinIOServiceStub(channel)")
        print()
        print("# è°ƒç”¨æœåŠ¡æ–¹æ³•")
        print("request = minio_service_pb2.HealthCheckRequest(detailed=True)")
        print("response = stub.HealthCheck(request)")
        print("print(f'å¥åº·çŠ¶æ€: {response.healthy}')")
        print()
        print("channel.close()")
        print("```")
        
    except KeyboardInterrupt:
        print("\n\nâš ï¸  æµ‹è¯•è¢«ç”¨æˆ·ä¸­æ–­")
        sys.exit(1)
    except Exception as e:
        print(f"\n\nâŒ æµ‹è¯•å¤±è´¥: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == '__main__':
    main()



